virtcontainers/fc.go

```diff
@@ -0,0 +1,666 @@
+	driveID := "drive-" + strconv.Itoa(drive.Index)
+	driveParams := ops.NewPatchGuestDriveByIDParams()
+	driveParams.SetDriveID(driveID)
+
+	driveFc := &models.PartialDrive{
+		DriveID:    &driveID,
+		PathOnHost: &drive.File, //This is the only property that can be modified
+	}
+	driveParams.SetBody(driveFc)
+	_, err := fc.client().Operations.PatchGuestDriveByID(driveParams)
+	if err != nil {
+		return err
+	}
+
+	// Rescan needs to used only if the VM is running
+	if fc.vmRunning() {
+		actionParams := ops.NewCreateSyncActionParams()
+		actionInfo := &models.InstanceActionInfo{
+			ActionType: "BlockDeviceRescan",
+			Payload:    driveID,
+		}
+		actionParams.SetInfo(actionInfo)
+		_, err = fc.client().Operations.CreateSyncAction(actionParams)
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// addDevice will add extra devices to firecracker.  Limited to configure before the
+// virtual machine starts.  Devices include drivers and network interfaces only.
+func (fc *firecracker) addDevice(devInfo interface{}, devType deviceType) error {
+	span, _ := fc.trace("addDevice")
+	defer span.Finish()
+
+	fc.state.RLock()
+	defer fc.state.RUnlock()
+
+	if fc.state.state == notReady {
+		dev := firecrackerDevice{
+			dev:     devInfo,
+			devType: devType,
+		}
+		fc.Logger().Info("FC not ready, queueing device")
+		fc.pendingDevices = append(fc.pendingDevices, dev)
+		return nil
+	}
+
+	fc.Logger().Info("adding a device")
+	switch v := devInfo.(type) {
+	case Endpoint:
+		fc.Logger().Info("Adding net device")
+		return fc.fcAddNetDevice(v)
+	case config.BlockDrive:
+		fc.Logger().Info("Adding block device")
+		return fc.fcAddBlockDrive(v)
+	case kataVSOCK:
+		fc.Logger().Info("Adding vsock")
+		return fc.fcAddVsock(v)
+	default:
+		fc.Logger().Warning("unknown device: %v", devInfo)
+		break
+	}
+
+	return nil
+}
+
+// hotplugAddDevice supported in Firecracker VMM
+func (fc *firecracker) hotplugAddDevice(devInfo interface{}, devType deviceType) (interface{}, error) {
+	span, _ := fc.trace("hotplugAddDevice")
+	defer span.Finish()
+
+	switch devType {
+	case blockDev:
+		//The drive placeholder has to exist prior to Update
+		return nil, fc.fcUpdateBlockDrive(*devInfo.(*config.BlockDrive))
+	default:
+		fc.Logger().WithFields(logrus.Fields{"devInfo": devInfo,
+			"deviceType": devType}).Warn("hotplugAddDevice: unknown device")
+		break
+	}
+
+	return nil, nil
+}
+
+// hotplugRemoveDevice supported in Firecracker VMM, but no-op
+func (fc *firecracker) hotplugRemoveDevice(devInfo interface{}, devType deviceType) (interface{}, error) {
+	return nil, nil
+}
+
+// getSandboxConsole builds the path of the console where we can read
+// logs coming from the sandbox.
+//
+// we can get logs from firecracker itself; WIP on enabling.  Who needs
+// logs when you're just hacking?
+func (fc *firecracker) getSandboxConsole(id string) (string, error) {
+	return "", nil
+}
+
+func (fc *firecracker) disconnect() {
+	fc.state.set(notReady)
```

Comment:
Why is fc state changed here? I thought the state should reflect the real state of the FC guest. And the `disconnect()` call only means to close the current runtime <-> hypervisor control channel connection, which is `fc.fcClient`.

Reply:
This is just a logical disconnect. We may still need the client connection for cleanup. We can change this later.